[OUTDATED] - Add ARS offramping support.

gianfra-t opened the pull request:
Issues: #173 and #172.

Besides implementing the basic Anclap integration, this PR adds a temporary endpoint for the `signer-service` to allow a single account controlled by Vortex to sign the `sep-10`, only valid for Anclap. This account must hold some ARS token before the off-ramp.

ebma commented on src/assets/coins/ARS.svg:
This doesn't look like ARS to me 😅

gianfra-t replied:
We are part of the EU now  😂. It slipped, I'll remove it.

ebma commented on src/services/anchor/index.ts:
```diff
@@ -194,10 +197,20 @@ export async function sep24First(sessionParams: IAnchorSessionParams): Promise<I
+
+  if (sessionParams.tokenConfig.stellarAsset.code.string === 'ARS\0') {
+    sep24Params = new URLSearchParams({
+      asset_code: sessionParams.tokenConfig.stellarAsset.code.string.replace('\0', ''),
```
Since we explicitly check for the ARS token here, I guess you can also just do
```suggestion
      asset_code: 'ARS',
```

gianfra-t replied:
True, good one.

ebma commented on src/services/anchor/index.ts:
```diff
@@ -194,10 +197,20 @@ export async function sep24First(sessionParams: IAnchorSessionParams): Promise<I
+    sep24Params = new URLSearchParams({
+      asset_code: sessionParams.tokenConfig.stellarAsset.code.string.replace('\0', ''),
+      amount: sessionParams.offrampAmount,
+      account: stellarPublic, // For Ancalap, we need to send this parameter.
```
```suggestion
      account: stellarPublic, // For Anclap, we need to send this parameter.
```

TorstenStueber commented on signer-service/src/constants/tokenConfig.js:
```diff
@@ -20,6 +20,14 @@ const TOKEN_CONFIG = {
+    vaultAccountId: '6bE2vjpLRkRNoVDqDtzokxE34QdSJC2fz7c87R9yCVFFDNWs',
+    minWithdrawalAmount: '10000000000000',
+    maximumSubsidyAmountRaw: '100000000000000', // 100 unit ~ 0.1 USD @ Oct/2024
+  },
```
What about the `pendulumCurrencyId`?

gianfra-t replied:
Was added locally, not yet pushed at the time. Pushed now.

TorstenStueber commented on src/constants/tokenConfig.ts:
```diff
@@ -89,6 +89,29 @@ export const OUTPUT_TOKEN_CONFIG: Record<OutputTokenType, OutputTokenDetails> =
+    },
+    vaultAccountId: '6bE2vjpLRkRNoVDqDtzokxE34QdSJC2fz7c87R9yCVFFDNWs',
+    erc20WrapperAddress: '6cNENXUqHUeEGSm4psQCeykZiLXJL9VzMQnvSoouyeEEoJpe',
+    minWithdrawalAmountRaw: '100000000000000', // 100 ARS?
```
The limits according to https://api.anclap.com/transfer24/info are 11 and 500.000

TorstenStueber reviewed (commented):
It seems that Anclap has a [fixed and a percentage fee together](https://api.anclap.com/transfer24/info) (as is permitted by SEP-24), which would mean that we need to general our fee calculation. Also the token configuration cannot just have a `offrampFeesBasisPoints` but that would need to have a fixed and a relative part now.

gianfra-t commented:
@pendulum-chain/devs I will address your requested changes, but as we discussed in this [thread](https://satoshipay.slack.com/archives/C01V1F56RMJ/p1730482546101089?thread_ts=1730311660.344149&cid=C01V1F56RMJ), we will probably never merge this PR, all these changes will be incorporated in the PR regarding the `memo` implementation for `sep-10` (issue [here](https://github.com/pendulum-chain/vortex/issues/237)).

ebma commented on signer-service/src/constants/constants.js:
```diff
@@ -15,6 +15,7 @@ require('dotenv').config();
 const PENDULUM_FUNDING_SEED = process.env.PENDULUM_FUNDING_SEED;
 const FUNDING_SECRET = process.env.FUNDING_SECRET;
 const MOONBEAM_EXECUTOR_PRIVATE_KEY = process.env.MOONBEAM_EXECUTOR_PRIVATE_KEY;
+const SEP10_MASTER_SECRET = process.env.SEP10_MASTER_SECRET;
```
Let's maybe set this to the same value so that it's easy to change back if we decide to have two separate accounts in the future?
```suggestion
const SEP10_MASTER_SECRET = FUNDING_SECRET;
```

gianfra-t commented on signer-service/src/api/routes/v1/stellar.route.js:
```diff
@@ -8,4 +8,9 @@ router.route('/create').post(validateCreationInput, controller.createStellarTran
 
 router.route('/payment').post(validateChangeOpInput, controller.changeOpTransaction);
 
+// TODO maybe add validator.
```
These are the temporary endpoints. `GET` will return the public key, required to start the `sep-10` on the first place. `POST` will provide the signature.

I will leave this comment about validators, but I don't think is of too much value to add one now, considering these are only temporary.

gianfra-t commented on signer-service/src/api/services/sep10.service.js:
```diff
@@ -0,0 +1,76 @@
+
+const NETWORK_PASSPHRASE = Networks.PUBLIC;
+
+exports.signSep10Challenge = async (challengeXDR) => {
```
This endpoint will receive the challenge, check the operations to ensure that only Anclap is valid, and that the to-be-authorized account is our master account.

It is of course not the safest.

gianfra-t commented on src/components/FeeCollapse/index.tsx:
```diff
@@ -6,9 +6,15 @@ import { useEventsContext } from '../../contexts/events';
+  const fees = toAmount.mul(feeBasisPoints).div(10000).add(fixedFees).round(2, 1);
+  const totalReceiveRaw = toAmount.minus(fees);
+
+  if (totalReceiveRaw.gt(0)) {
```
Correct for fixed fee from Anclap.

TorstenStueber commented on src/services/anchor/index.ts:
```diff
@@ -100,8 +113,16 @@ export const sep10 = async (
+  } else {
+    const { masterSignature, masterPublic } = await fetchMasterSignatureSep10(transactionSigned.toXDR());
+    transactionSigned.addSignature(masterPublic, masterSignature);
+    maybeMasterPublic = masterPublic;
```
Shouldn't this always be the same public key as the `masterSep10Public` fetched above anyway? Can we reuse it to make it more clear that this must be the same key?

gianfra-t replied:
Actually, good catch. No need to return it along with the signature, we can use the same as the one above.

TorstenStueber commented on src/services/anchor/index.ts:
```diff
@@ -194,10 +218,23 @@ export async function sep24First(sessionParams: IAnchorSessionParams): Promise<I
-  });
+  let sep24Params;
+
+  if (sessionParams.tokenConfig.stellarAsset.code.string === 'ARS\0') {
```
Let's also rather use a flag in the token config instead of a hard coded condition here.

TorstenStueber commented on src/services/anchor/index.ts:
```diff
@@ -194,10 +218,23 @@ export async function sep24First(sessionParams: IAnchorSessionParams): Promise<I
+    sep24Params = new URLSearchParams({
+      asset_code: 'ARS',
+      amount: sessionParams.offrampAmount,
+      account: masterPublic, // Since we signed with the master from the service, we need to specify the corresponding public here
```
Shouldn't this be the ephemeral account instead? The SEP-24 standard defines `account` as:

> The Stellar or muxed account the client will use as the source of the withdrawal payment to the anchor. Defaults to the account authenticated via SEP-10 if not specified.

gianfra-t replied:
Yes, you are right. But it doesn't work with it. It only works with the one that did the `sep-10`. I missed that part of the docs and found this by trial an error.

Sharqiewicz commented on signer-service/src/api/services/sep10.service.js:
```diff
@@ -0,0 +1,76 @@
+  let hasWebAuthDomain = false;
+
+  // Verify extra manage_data operations
+  for (let i = 1; i < operations.length; i++) {
```
Why not use operations.map() instead of a for loop?

gianfra-t replied:
I found the explicit syntax more clear for this case such that it's visible what are we filtering for, but we could add a generic `verifyOperation` with all the rules and map over that, WDYT?
And return the first error it encounters.

Sharqiewicz replied:
I think writing a function with a clear name and adding it to .map() can make the code more readable and self-documenting. Using a function `verifyOperation` with `.map` makes it easier to follow what is happening. wdyt?

gianfra-t replied:
The issue is that it will require quite a lot of context and although it is easier to read, the control flow becomes harder as well as the errors returns. But let's see, I'll try to see how it looks.

Sharqiewicz commented on src/services/anchor/index.ts:
```diff
@@ -77,9 +79,20 @@ export const sep10 = async (
-  });
+
+  let urlParams;
+  if (outputToken === 'ars') {
```
I think refactoring this to a function and then `const urlParams = theFunction()` would improve readability and reusability

gianfra-t replied:
Alright! I will add it also including this [comment](https://github.com/pendulum-chain/vortex/pull/201#discussion_r1829797828).

gianfra-t commented on signer-service/src/constants/tokenConfig.js:
```diff
@@ -20,6 +20,22 @@ const TOKEN_CONFIG = {
+    assetCode: 'ARS\0',
+    assetIssuer: 'GCYE7C77EB5AWAA25R5XMWNI2EDOKTTFTTPZKM2SR5DI4B4WFD52DARS',
+    vaultAccountId: '6bE2vjpLRkRNoVDqDtzokxE34QdSJC2fz7c87R9yCVFFDNWs',
+    minWithdrawalAmount: '10000000000000',
```
You're right. I will just leave it to `11` for consistency. We may need it soon.

gianfra-t commented on signer-service/src/api/services/sep10.service.js:
```diff
@@ -0,0 +1,76 @@
+
+  // Purposely hardcode Anclap's key, we don't want to sign this type of transaction
+  // for Mykobo
+  const expectedKey = `api.anclap.com auth`;
```
To make it generic for future sep-10 with memos, you mean? Sure, I hard coded it to save time, but I can make it generic.

gianfra-t commented on src/services/anchor/index.ts:
```diff
@@ -77,9 +79,20 @@ export const sep10 = async (
-  });
+
+  let urlParams;
+  if (outputToken === 'ars') {
```
Yes, totally.

Sharqiewicz commented on signer-service/src/api/services/sep10.service.js:
```diff
@@ -0,0 +1,76 @@
+  }
+
+  // Verify manage_data operations
+  const operations = transactionSigned.operations;
```
wdyt?

Sharqiewicz commented on src/services/stellar/index.tsx:
```diff
@@ -188,7 +191,7 @@ async function createOfframpAndMergeTransaction(
   }
 
-  const stellarAsset = new Asset(code.string, issuer.stellarEncoding);
+  const stellarAsset = new Asset(code.string.replace('\0', ''), issuer.stellarEncoding);
```
If we're using string.replace('\0', '') or assetCode.replace('\0', '') in multiple places, perhaps we should extract it into a separate function. This would enhance both readability and reusability. What do you think?

gianfra-t commented on src/services/anchor/index.ts:
```diff
@@ -100,8 +113,16 @@ export const sep10 = async (
-  transactionSigned.sign(ephemeralKeys);
+  // only sign if not ARS token, else call the temporary
+  // signing service's endpoint
+  let maybeMasterPublic;
```
Yup! Improved [here](https://github.com/pendulum-chain/vortex/pull/201/commits/29a9147614fb8e8e01f253d0d490292bc209f2f9#diff-e8e6c9ebebea2e994a257e9f9c83bb19da16efb95e46255edfd5837a1f2a2f99R113). The `sep10Account` variable now holds either the ephemeral or the "client master".

gianfra-t commented on signer-service/src/api/helpers/anchors.js:
```diff
@@ -0,0 +1,27 @@
+  }
+
+  const tomlFileContent = (await response.text()).split('\n');
+  const findValueInToml = (key) => {
```
I think it's simplified now!

gianfra-t commented:
@pendulum-chain/devs as we discussed in the sync and on slack, this solution (signing with a fixed/master account always) doesn't work anymore and will probably stay that way. Nevertheless, many of the changes (like adding ARS icon, config, toml, etc) will remain.  
This [PR](https://github.com/pendulum-chain/vortex/pull/236) will be merged and has included the integration (with ephemeral) and many of the requested changes.

gianfra-t commented:
I will close this PR to avoid confusion. TL;DR:
- The solution of a single account is incorrect/incomplete.
- The remaining part of the integration was done [here](https://github.com/pendulum-chain/vortex/pull/236), which also addresses the comments.

gianfra-t commented:
I will close this PR to avoid confusion. TL;DR:
- The solution of a single account is incorrect/incomplete.
- The remaining part of the integration was done [here](https://github.com/pendulum-chain/vortex/pull/236), which also addresses the comments.
